### PR TITLE
[REG-1342] Implementation of features to support Validations MVP

### PR DESCRIPTION
**Pull request by vontell**
This PR now addresses REG-1342, which is all about supporting validations, screenshots, and replay data for local bots.
You can try this out in BossRoom by doing the following:

* Pulling this branch
* Pull the backend code from here, and restart RGService - https://github.com/Regression-Games/RegressionGames/pull/430
* Make a local bot from the code at the end of this description, and then sync your bots.

Once you run your bot in bossroom, you should be able to go to the bot history page and open validations! Make sure your client is up to date as well.

Tasks remaining for this ticket:

* Review the threading and async approach and fix any issues you think may be there
* There is a bug with replay data - I _think_ the format I am saving is valid, but the replay data pane won't load the zip, due to the RGEntity interface - we should verify that remote replay data still works

```csharp
using System;
using System.Collections.Generic;
using RegressionGames;
using RegressionGames.RGBotLocalRuntime;
using RegressionGames.StateActionTypes;
using UnityEngine;

namespace AbilityBot_0
{
    public class BotEntryPoint : RGUserBot
    {

        protected override bool GetIsSpawnable()
        {
            return true;
        }

        protected override RGBotLifecycle GetBotLifecycle()
        {
            return RGBotLifecycle.MANAGED;
        }

        public override void ConfigureBot(RG rgObject)
        {
            var classIndex = new System.Random().Next(4);
            var classes = new string[] {"Mage", "Rogue", "Tank", "Archer"};
            rgObject.CharacterConfig = new Dictionary<string, object>() {
                { "characterType", $"{classes[classIndex]}" }
            };
        }

        public override void ProcessTick(RG rgObject)
        {
            try
            {
                var entities = rgObject.FindEntities();
                if (entities.Count > 0)
                {
                    // default to friendly ability
                    var skillId = 0;
                    var target = entities[new System.Random().Next(entities.Count)];
                    if (target.GetField<int>("team", 0) == 1)
                    {
                        // if enemy.. set ability to 1
                        skillId = 1;
                    }

                    rgObject.RecordValidation("Test passed", RGValidationResultType.PASS, "heart");
                    rgObject.RecordValidation("Test failed", RGValidationResultType.FAIL, "heart");
                    rgObject.RecordValidation("Test warning", RGValidationResultType.WARNING, "heart");

                    var targetPosition = target.position;
                    var action = new RGActionRequest("PerformSkill", new Dictionary<string, object>()
                    {
                        { "skillId",  skillId},
                        { "targetId", target["id"] },
                        { "xPosition", targetPosition.x },
                        { "yPosition", targetPosition.y },
                        { "zPosition", targetPosition.z },
                    });
                    rgObject.PerformAction(action);
                }
                else
                {
                    RGDebug.LogWarning("No entities found...");
                }
            }
            catch (Exception ex)
            {
                RGDebug.LogError($"Error getting target position: {ex}");
            }
        }
    }
}
```

### CODE REVIEW

**Comment by analogrelay** [comment]
> NOTE: I am making this PR so @addisonbgross you can run new validation tests locally and get a real request working. Please see the linear ticket on my instructions on how to do that locally.

May be worth making this a Draft PR unless you're ready for folks to review and provide feedback. My usual interpretation of a non-draft PR is that it's ready for review, but your PR description indicates to me that it's not ready to be reviewed yet.

**Comment by RG-nAmKcAz** [comment]
> > NOTE: I am making this PR so @addisonbgross you can run new validation tests locally and get a real request working. Please see the linear ticket on my instructions on how to do that locally.
> 
> May be worth making this a Draft PR unless you're ready for folks to review and provide feedback. My usual interpretation of a non-draft PR is that it's ready for review, but your PR description indicates to me that it's not ready to be reviewed yet.

Seconded / Amen

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (addisonbgross) [inline Runtime/Scripts/DataCollection/RGDataCollection.cs:102]: This might be good to wrap in a `try/catch`. If an upload fails, we can still try the others

Comment (addisonbgross) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs:102]: Nevermind. It looks like the `UploadScreenshot` function handles that. I'll leave this comment here for fun

**Comment by vontell** [comment]
@RG-nAmKcAz @analogrelay my apologies, I forgot to remove that when I requested reviews on this. Will fix!

File: Runtime/Scripts/RGBotServerListener.cs
Comment (vontell) [inline Runtime/Scripts/RGBotServerListener.cs:287]: Any ideas here? Because the upload functionality is async, this gets called before RecordSession is finished. Maybe I'll try to add a wait for tasks call in the RecordSession function somehow.

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (addisonbgross) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs:47]: I couldn't find anything about how to solve this situation :( 
I think mentioning that 1x scale is needed for the MVP is a solid choice

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (analogrelay) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs:60]: 1. There's no guarantee that `TcpClient.Close` has finished closing the connection when it returns, so `RGClientConnection_Remote.Close` also can't provide that guarantee.
2. `UploadScreenshot` triggers a `UnityWebRequest`, which is an async operation but doesn't `await` it. So it will return from this method before the upload has completed.

Async operations **that actually use a background thread** like `UnityWebRequest` work by kicking off the request in the background and polling for status on each frame. If you `await` the `Task` returned by `UploadScreenshot`, execution of the rest of this method is deferred until the next frame _after the request completes_

File: Runtime/Scripts/RGBotServerListener.cs
Comment (analogrelay) [reply Runtime/Scripts/RGBotServerListener.cs:287]: If you make `RecordSession` an `async Task` method, you can `await` that `Task`. That will ensure the rest of the method doesn't execute until all the async operations `await`ed in `RecordSession` complete. It'll mean the execution of this method will spread over several frames, but if you want to wait for the screenshots to upload, that's what you need to do.

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (RG-nAmKcAz) [inline Runtime/Scripts/DataCollection/RGDataCollection.cs:34]: These should all be using RGDebug at appropriate log levels for each message

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (RG-nAmKcAz) [inline Runtime/Scripts/DataCollection/RGDataCollection.cs:77]: should separate `$"RGData/{_sessionName}"` into `"RGData", _sessionName` so that the path helper is building it OS agnostic

File: Runtime/Scripts/RGBotServerListener.cs
Comment (RG-nAmKcAz) [inline Runtime/Scripts/RGBotServerListener.cs:63]: `RGDataCollection` not being a scriptableObject or MonoBehaviour made me do a double take as to why this was happening in `Awake`.  I would suggest moving this to `Start` (Awake can be called multiple times in a single game run instance) or just initializing it on the field itself which would be cleaner

**Review by RG-nAmKcAz — CHANGES_REQUESTED** [review]
A few small things in addition to @analogrelay's existing comments.

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (analogrelay) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs:77]: Good catch, missed that one 🤣

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (rcornwall-reg) [inline Runtime/Scripts/DataCollection/RGDataCollection.cs:51]: Also expanding on what Zack said, but it should be:
```
RGDebug.Log("Ending data collection, uploading data to Regression Games...");
```

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (rcornwall-reg) [inline Runtime/Scripts/DataCollection/RGDataCollection.cs:89]:
```
RGDebug.Log("Deleted all local temp data collected for RG bots");
```

**Review by rcornwall-reg — CHANGES_REQUESTED** [review]

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (vontell) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs:47]: Thanks for this! Yeah def agree we can do that later, I'll make sure that once we write up the docs for the validation work, I'll make a note that only 1x scale is supported in the MVP.

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (vontell) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs]: Good catch, fixed!

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (vontell) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs]: Good call, added!

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (vontell) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs]: Fixed!

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (vontell) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs]: The reason I did this was because Join only supports up to 3 path pieces (very strange) - updated this to use Combine which does supports a variable number of path pieces, and works as long as the first path is an absolute path (which it is in this case).

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (vontell) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs]: Good suggestion! Fixed.

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (vontell) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs:94]: Fixed!

File: Runtime/Scripts/RGBotServerListener.cs
Comment (vontell) [reply Runtime/Scripts/RGBotServerListener.cs]: Moved to start - also, it now doesn't even need a reference to the gameobject.

**Review by vontell — COMMENTED** [review]
Addressed all of the comments so far **except** the async ones, gonna do that soon...

File: Runtime/Scripts/RGBotServerListener.cs
Comment (RG-nAmKcAz) [reply Runtime/Scripts/RGBotServerListener.cs]: initializing it on the field itself would have been cleaner, but closing this comment

**Review by RG-nAmKcAz — APPROVED** [review]
approved pending async fixes

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (RG-nAmKcAz) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs:63]: In addition to @analogrelay's analysis....

The core issue you have @vontell is with where you are calling `RecordSession` in `RGBotServerListener.EndClientConnection`

For `REMOTE` bots only, you need to call `RecordSession` in the error/success handler of the call to `RGService` to end the client connection. <- This call is what creates the bot history record.

Something akin to this...
```csharp
// --- Don't call this for local bots ...
if (clientConnection.Type == RGClientConnectionType.REMOTE)
{
    _ = RGServiceManager.GetInstance()?.StopBotInstance(
        clientId,
        () => {
            // Upload the replay data for this bot
            _dataCollection.RecordSession(clientId, clientConnection.Type);
        },
        () => {
            // Upload the replay data for this bot
            _dataCollection.RecordSession(clientId, clientConnection.Type);
        }
    );
} else 
{
    // Upload the replay data for this bot
    _dataCollection.RecordSession(clientId, clientConnection.Type);
}
```

**Comment by vontell** [comment]
Changing this to draft again as I push some in-progress code

File: Runtime/Scripts/RGBotServerListener.cs
Comment (vontell) [inline Runtime/Scripts/RGBotServerListener.cs:785]: @RG-nAmKcAz you probs have some insights into this - I see that our serialization for `RGTickInfoData` is a little weird. When I save this data after converting to JSON, I get the tick and scene name, but no state. Any ideas how I make sure that the entire object with the game state gets serialized?

```json
{"actions":[{"action":"PerformSkill"}],"validationResults":[{"name":"Test passed","result":0,"icon":"heart","id":"54ae616b-376c-46af-995b-1c9dae19fdfb","tick":950,"message":"Test passed","passed":true,"version":1},{"name":"Test failed","result":1,"icon":"heart","id":"d752c8e2-1c3c-4a2e-bef7-a295386f4269","tick":950,"message":"Test failed","passed":false,"version":1},{"name":"Test warning","result":2,"icon":"heart","id":"d149ce27-070d-4f38-a6a6-5b23c6d8ca96","tick":950,"message":"Test warning","passed":false,"version":1}],"error":"","tickInfo":{"tick":950,"sceneName":"BossRoom"}}
```

Comment (vontell) [reply Runtime/Scripts/RGBotServerListener.cs:785]: Figured I'd ask if you had a quick answer - otherwise, I'll implement a serialization function that uses the RGTickInfoData ToSerialize method as part of it potentially, or at least following that pattern.

Comment (RG-nAmKcAz) [reply Runtime/Scripts/RGBotServerListener.cs:785]: we call `.ToString()` on this explicitly when sending remote socket data is how we're working today.

haven't dug into your code to see how you're trying to serialize the object, make sure first that you're using `JsonConvert`, if its still not working with that, then yes I'd write your own serialize method

Comment (vontell) [reply Runtime/Scripts/RGBotServerListener.cs:785]: Your suggestions saved me! I was using the wrong JSON conversion utility, and then I ran into a looping reference issue so created a serialization method to address that. Thanks!

**Comment by vontell** [comment]
@ reviewers this is finally ready for review again, and has changed a lot since the last review. See the description for updates on where this is currently.

Also relies on https://github.com/Regression-Games/RegressionGames/pull/430

**Review by RG-nAmKcAz — CHANGES_REQUESTED** [review]

**Comment by vontell** [comment]
Thanks for the comments @RG-nAmKcAz - for reviewers, any reviewer who has some free time, I will need someone to address those for me.

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (vontell) [inline Runtime/Scripts/DataCollection/RGDataCollection.cs:94]:
```suggestion
                    // if we already took that screen shot, grab the next number from the queue
````

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (vontell) [inline Runtime/Scripts/DataCollection/RGDataCollection.cs:116]: This seems better! Thanks!

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (vontell) [inline Runtime/Scripts/DataCollection/RGDataCollection.cs:142]: I see, this makes more sense now

**Review by RG-nAmKcAz — APPROVED** [review]

**Comment by RG-nAmKcAz** [comment]
@analogrelay This is now good to go and @vontell has approved the changes I made today to finish it up.

Let me know if you see anything remaining you want us to fix before merging.

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (RG-nAmKcAz) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs]: confirmed that Aaron had made these updates ... resolving this as ashley is out for a few days

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (RG-nAmKcAz) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs]: confirmed that Aaron had made these updates ... resolving this as ashley is out for a few days

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (RG-nAmKcAz) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs]: There is a lot of use of "/" as a convention in this file still.  I think we're ok as Windows / Linux / Mac generally all accept that format

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (RG-nAmKcAz) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs]: confirmed that Aaron had made these updates ... resolving this as ashley is out for a few days

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (RG-nAmKcAz) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs]: confirmed that Aaron had made these updates ... resolving this as ashley is out for a few days

File: Runtime/Scripts/RGBotServerListener.cs
Comment (RG-nAmKcAz) [reply Runtime/Scripts/RGBotServerListener.cs]: confirmed that Aaron had made these updates ... resolving this as ashley is out for a few days

**Comment by RG-nAmKcAz** [comment]
@rcornwall-reg  This is now good to go and @vontell has approved the changes I made today to finish it up.

Let me know if you see anything remaining or giveth your approval once you are satisfied that your requested changes have been made.

**Review by rcornwall-reg — APPROVED** [review]
LG! This might be it!

**Pull request merged**
